### DISCUSSION TITLE
Usage of the 'Session.add()' during flush should be an exception and not a warning

### DISCUSSION THREAD

**Discussion by Zaczero** [sqlalchemy/sqlalchemy#11599] [Ideas]
[This warning](https://github.com/sqlalchemy/sqlalchemy/blob/96f1172812f858fead45cdc7874abac76f45b339/test/orm/test_utils.py#L133), which is triggered [here](https://github.com/sqlalchemy/sqlalchemy/blob/96f1172812f858fead45cdc7874abac76f45b339/lib/sqlalchemy/orm/session.py#L3484), should be an exception and not a warning. As the warning says "Results may not be consistent". This can and will lead to silent data corruption and it's hopeful to assume a warning is enough.

**Comment by zzzeek** [comment]
the unit of work in modern releases runs flush repeatedly until all pending changes in the Session are exhausted, so the general pattern of making changes within events like after_insert likely "works" now, however we just didnt want to have to support complex session manipulations within those hooks working perfectly afterwards.

It's likely that session manipulation inside these event hooks does actually work, I would have to be creative to think of a case where it doesnt; the "flush repeating" feature which was added to support the after_flush() hook might actually handle all of it.


demo

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import object_session
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

val = False

@event.listens_for(A, "before_insert")
def evt(mapper, connection, target):
    global val
    if not val:
        object_session(target).add(A(data="before insert"))
        val = True


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(data="main"))
s.commit()

```

**Reply by Zaczero** [reply]
Hmm.. I have only noticed this issue because one of my tests detected some updates not being pushed to the database (without errors). It happened very rarely indicating some race condition and both times the same SAWarning was present in the logs. Perhaps there is a bug in the retry mechanism? [This line](https://github.com/openstreetmap-ng/openstreetmap-ng/blob/67cef7887e6de74f4ce0139194363e0fb24059cd/app/services/optimistic_diff/apply.py#L108) was not being pushed to the database, when [this line](https://github.com/openstreetmap-ng/openstreetmap-ng/blob/67cef7887e6de74f4ce0139194363e0fb24059cd/app/services/optimistic_diff/apply.py#L190) was flushing other updates.

[Job 1](https://github.com/openstreetmap-ng/openstreetmap-ng/actions/runs/9880466795/job/27289067899)
[Job 2](https://github.com/openstreetmap-ng/openstreetmap-ng/actions/runs/9896428758/job/27338623915)

Should I open a bug report?

**Reply by zzzeek** [reply]
if you can make a test case for me that shows it not working I will have a better idea of this scenario.

it's not that controversial to make this warning an error anyway, it would a be a 2.1 thing though

**Reply by Zaczero** [reply]
Hey! I didn’t have enough time to prepare an example, but I decided to take a quick look at the source code. As far as I can see, there is no multi-thread synchronization between `flush` and `any_state_modifier` at all. The same applies to `any_state_modifier` and `any_state_modifier`. This library does not seem to be thread-safe, so running multiple awaits/threads at the same time is all about luck.

This could be resolved in one of three ways:
- Implement proper async methods instead of thread wrappers: may not be feasible and will not work for Python 3.14 NoGIL.
- Add synchronization primitives: very difficult to get right at scale.
- Raise exceptions for safety when run in parallel: not a proper fix, but it will avoid potential issues.

**Reply by CaselIT** [reply]
That's documented that session is not concurrent safe. Note that it sqlalchemy used no thread wrapper.

In general it doesn't make sense to concurrently use a db connection, since the db generally do one operation at a time on a single connection

**Reply by Zaczero** [reply]
> Note that it sqlalchemy used no thread wrapper.

That's right, my bad.

> In contrast, greenlets are cooperative and sequential. This means that when one greenlet is running, no other greenlet can be running; ([source](https://greenlet.readthedocs.io/en/latest/))

Perhaps because .flush() is a greenlet, but .add() is not, it introduces the race condition?

**Reply by CaselIT** [reply]
There is no race conditions knowns if you use the session in a non concurrent way, that as mentioned is not a supported usage.

If you believe to have found one please post an example of the code that's causing it

**Reply by Zaczero** [reply]
Thanks for the clarification! You are right. So, this discussion boils down to changing the warning to an exception if such behavior is not supported. Data consistency should be the priority in cases like these. In my case, I didn’t realize I was running concurrent SQLAlchemy code since `.add()` is not async. Because `.flush()` is a greenlet (similar to a thread) wrapper, this leads to race conditions between `async`⭤`non-async` methods which I can now see. Thanks! 👍

> In general it doesn't make sense to concurrently use a db connection

For additional context, I was adding more data to the session while waiting for the flush in the background. Due to the volume of data, I wanted to start flushing some data early to speed up the overall process. I simply didn’t expect the non-async `.add()` to conflict with it.

**Reply by CaselIT** [reply]
I'm not clear how that would be possible. Something like this never causes a race condition
```
session.add(...)
...
await session.flush()
```

To have a race condition here you would have to do something like
```py
session.add(...)
...
coro = session.flush()
session.add(...)
...
await coro
```

But that's not generally how you write async code.

Also note that greenlet here plays no role at all, you would have the same issue without the use of greenlet

**Reply by Zaczero** [reply]
See https://github.com/sqlalchemy/sqlalchemy/discussions/11599#discussioncomment-10032031. I am using a [TaskGroup](https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup).

**Reply by CaselIT** [reply]
That like gather. You can't share the same session between multiple concurrent tasks

**Reply by Zaczero** [reply]
Thanks for the clarification! You are right. So, this discussion boils down to changing the warning to an exception if such behavior is not supported. Data consistency should be the priority in cases like these. In my case, I didn’t realize I was running concurrent SQLAlchemy code since `.add()` is not async. Because `.flush()` is a greenlet (similar to a thread) wrapper, this leads to race conditions between `async`⭤`non-async` methods which I can now see. Thanks! 👍
